dataset/bins.cpp

```diff
@@ -230,8 +244,29 @@ Variable concatenate_untyped(const VariableConstView &var, const Dim dim,
     return concatenate_typed<Dataset>(var, dim, shape, mask);
 }
 
+template <class T>
+void reserve_impl(const VariableView &var, const VariableConstView &shape) {
+  // TODO this only reserves in the bins, but assumes buffer has enough space
+  const auto &[indices, dim, buffer] = var.constituents<bucket<T>>();
+  variable::transform_in_place(
+      indices, shape,
+      overloaded{
+          core::element::arg_list<std::tuple<scipp::index_pair, scipp::index>>,
+          core::keep_unit,
+          [](auto &begin_end, auto &size) { begin_end.second += size; }});
```

Comment:
Is there some handing required to here to ensure that the new `begin_end.second` does not overlap with the next `begin_end.first` ?

Reply:
Yes, see TODO in line 249. This function is part of a set of functions making up the basic toolkit of working with bins, including others for compying, resizing, ... At this point I am very unhappy about this toolkit and would like to make improvement, so this might be removed or replaced soon (e.g., by a `reserve` that actually resizes the buffer?). I thus took some shortcuts in the implementation.